price-feeder/oracle/provider/websocket_controller.go

```diff
@@ -0,0 +1,189 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+)
+
+type (
+	MessageHandler func(int, []byte)
+
+	// WebsocketController defines a provider agnostic websocket handler
+	// that manages reconnecting, subscribing, and receiving messages
+	WebsocketController struct {
+		ctx              context.Context
+		providerName     Name
+		websocketURL     url.URL
+		subscriptionMsgs []interface{}
+		messageHandler   MessageHandler
+		logger           zerolog.Logger
+		client           *websocket.Conn
+		mu               sync.Mutex
```

Comment:
nit:
```suggestion
		mtx               sync.Mutex
```

Also, what's the advantage of using `sync.Mutex` here over `sync.RWMutex`?

Reply:
To answer the question: "why not just always use a RWMutex" some googling tells me the basic Mutex is faster when you don't need the read lock functionality.